### ISSUE TITLE
[formatting2] Performance issues with autowrap and already formatted document

### ISSUE THREAD

**Issue by mx990** [eclipse/xtext#3119]
There are performance issues when formatting an already properly formatted document in conjunction with autowrap.

When autowrap was enabled once in the current line, `FormattableDocument.createReplacements(...)` then checks for each replacer if it `needsAutowrap(...)`. This calls `ITextReplacerContext.getLeadingCharsInLineCount()` to determine if the line has exceeded the maximum line width.

`TextReplacerContext.getLeadingCharsInLineCount()` then iterates over all previous contexts until it finds a newline character in the respective replacements. However, if there are no replacements made because the document is already properly formatted, this iterates over all previous contexts until the beginning of the document.

There are also many replacer contexts created using `TextReplacerContext.withDocument(...)`, `withIndentation(...)`, and `withReplacer(...)`, where the resulting context is identical to the current one.

In summary, once autowrap is enabled, this may iterate over all replacer contexts for each formatter in the current line. This can be millions of iterations in total for a large document.
